### DISCUSSION TITLE
Blazor Hybrid app via BlazorwWebView in Avalonia

### DISCUSSION THREAD

**Discussion by qrzychu** [AvaloniaUI/Avalonia#8414] [Ideas]
It would be extremely good to be able to write Avalonia Blazor Hybrid aps, just like it is possible with MAUI, WPF and even Winforms

*What is Hybrid app? *

[Ms Docs](https://docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/?view=aspnetcore-6.0)

In general, it is a native app with a specialized web view to handle Blazor. Blazor components are run directly in process (no WASM, no JS), and the view just renders them. I don't know any technical details how to do it.

That way you get great performance (in process Blazor components), low memory usage (the only web layer is rendering), Blazor productivty and cross-platform support due to Avalonia.

I have no idea if this is hard task, but I think many people would prefer to host the Hybrid Blazor app in Avalonia over .NET MAUI if they go cross-platform.

**Comment by maxkatz6** [comment]
Yes, it's possible to implement. Furthermore, it should already work with no issues on Windows, where you can combine our NativeControlHost with WPF/WinForms BlazorWebView component. Similarly how I did it for the usual WebView component in Avalonia using native browser view - https://github.com/maxkatz6/AvaloniaWebView.

It's definitely possible to port other platforms in similar way, like they did for MAUI https://github.com/dotnet/maui/pull/654. Especially when Avalonia now supports native controls for mobile platforms in nightly builds, so it's possible to use native webkit for both android and ios.

Saying that, it's should be an interesting project, but it can't fit in the scope of our current release. And I am not sure if it's something that should be available as first class control in the Avalonia repository. Saying that, if community wants to implement that and maintain as an third party library, we can help with information how it should be implemented.

**Reply by maxkatz6** [reply]
Although, I would say from the beginning, native web view (with or without blazor) on the Linux will be a problematic feature. You would need to require every app to ship GTK-WebView most likely. And native embedding won't work on Wayland based desktop managers as for now, so it's hardly possible there.

**Reply by qrzychu** [reply]
@maxkatz6  I found this project: https://github.com/tryphotino/photino.Blazor

It's code isn't that long, maybe that would be a good start?

**Reply by maxkatz6** [reply]
It uses "Photino.NET", library to open a native web view in a separated window. Not exactly what you need if you want to embed it inside of existing Avalonia window.
But other parts of the code can be useful, like http handler or sync context.

**Reply by jpmikkers** [reply]
I just managed to create an Avalonia BlazorWebView component (based on the winforms BlazorWebView), together with a Avalonia-Blazor hybrid demo app. 

See here: [Baksteen.Avalonia.Blazor](https://github.com/jpmikkers/Baksteen.Avalonia.Blazor)

**Reply by qrzychu** [reply]
That's awesome! Thanks a lot :)

**Reply by danroth27** [reply]
@jpmikkers Looks like Baksteen.Avalonia.Blazor is Windows only right now. Any plans to make it xplat?

**Reply by jpmikkers** [reply]
@danroth27 Yes that's the goal, currently looking into the magic sauce behind BlazorWebView (as compared to WebView). After that I can maybe replace the Winforms WebView with CefSharp or another xplat browser engine.

**Reply by somelinguist** [reply]
Looks great! A few months ago I took a try at getting it to work with Eto.Forms, which can target  WPF, WinForms, MacOS, etc. 

For the WPF and WinForms targets, I just wrapped the native controls, as you have. 

For the MacOS one, I ended up spending a lot of time study the MAUI iOS source for BlazorWebView (and somewhat at Photino.Blazor's source as well), and I was able to get it working for Eto.Forms when targeting net6.0-macos (Eto.Forms has a few other options for targeting MacOS as well).

I haven't done much with it since then, but if it helps, the source is here: https://github.com/somelinguist/Eto.Blazor

**Reply by jpmikkers** [reply]
@somelinguist that's going to help a lot!! I see I may have to switch my license :)

**Reply by Mrxx99** [reply]
@jpmikkers Maybe also this can help: https://github.com/tryphotino/photino.Blazor

**Reply by somelinguist** [reply]
@jpmikkers Glad it can help! I haven't had much time to work more on the Eto.Forms control.

The main thing is configuring whatever webview component your using and getting it hooked up to a a custom subclass of `WebViewManager` https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.webview.webviewmanager?view=aspnetcore-7.0.

The WinForms BlazorWebView component implements this for you with WebView2.

My Mac handler for Eto.Forms wires up a WebViewManager subclass with WkWebview, which is available in net6.0-macOS targets. 

Photino.Blazor wires up their WebViewManager subclass with their implementation of a cross-platform web window that wraps the native webview for each system.

For Avalonia, you'll need to decide what webview(s) to wrap since there isn't (last I checked) an included webview control. And for Mac, for example, I don't think you have access to WkWebView unless you target net6.0-macos, net7.0-macos or Xamarin.Mac

Also, if I remember correctly, because of the way the base WebViewManager source is written, you might end up having to re-implement a lot of the other needed components like RootComponent, etc to use them in a cross-platform way between the different subclasses of WebViewManager. If I remember right, you only get public access to some of the official ones when you open platform-specific namespaces (or MAUI specific namespaces) in `Microsoft.AspNetCore.Components.WebView` and a few others.

It was definitely a learning experience for me, but it was fun. It's been a while since I looked at it though. :)

Best of luck!

**Reply by jpmikkers** [reply]
@somelinguist I'm at the point now (wip branch) where I have my own working reimplementation of BlazorWebView, indeed I had to open up various things that were marked internal. It still uses the Winforms WebView, but at least I have a working starting point and 'sandbox' now to replace that by something else.

**Reply by somelinguist** [reply]
@jpmikkers Thanks for pointing out your WIP branch!

That definitely sounds like a good way to go, especially if you can hook up a cross-platform webview or an abstraction around them. That’s basically what Photino does.

For Eto.Forms, since I could just plug in both the WPF and WinForms implementations, I started there. Then I figured out more how it actually works (sort of) for the Mac implementation. I mostly just ported MAUI’s iOS parts of it. 

While trying to figure out how to add GTK support (since Eto can target that too), I thought of trying to rewrite it to be more abstract, similar to Photino. But the Eto WebView control doesn’t provide enough access directly to the underlying platform webviews to do what’s needed via the basic control. I started looking into extending it, but I’m still not sure if I could get it to work for GTK targets yet.

I’ll look forward to see what you’re doing with Avalonia!

**Reply by jpmikkers** [reply]
@somelinguist how on earth did you figure out the JavaScript to connect the sendmessage/receivemessage?

**Reply by somelinguist** [reply]
@jpmikkers for Mac, I just copied it from the Maui iOS implementation here:

https://github.com/dotnet/maui/blob/3859318f971b39b9d4a0e48492a3daed012cb695/src/BlazorWebView/src/Maui/iOS/BlazorWebViewHandler.iOS.cs#L27-L50

I don't thing I remember changing anything for that part. :)

Many of the changes from the Maui iOS version were just changing UI prefixes to NS, and then making sure it interacted with Eto.Forms handlers.

**Reply by jpmikkers** [reply]
Well, I have CefSharp based BlazorWebView up and running (code: [see here](https://github.com/jpmikkers/Baksteen.Avalonia.Blazor/tree/wip/Baksteen.Blazor.CefSharpWinForms)). 
Next step is to use CefGlue instead, it should be able to run in Linux.

**Reply by idan-h** [reply]
Any update or help needed with that? @jpmikkers

**Reply by jpmikkers** [reply]
Waiting for a decent x-platform open source c# WebView. Last time I tried CefGlue I couldn't get it to run in-process.

**Reply by jairbubbles** [reply]
@jpmikkers  What about https://github.com/webview/webview_csharp?

**Reply by mondermurshed** [reply]
@jpmikkers
Hey, what is the status of the library? 
I was looking to build the UI using Blazor inside Avalonia shell, so is the project done yet for production purposes? 

My main goal is to build for Windows, MacOS and Linux, looking forward for your answer.

**Reply by jpmikkers** [reply]
@mondermurshed my project is still windows only unfortunately. I'm keeping an eye on photino.net for cross platform Blazor hybrid projects.

**Reply by mondermurshed** [reply]
All the support for you for trying to provide what the Avalonia team didn't for years.  

For now i will just use WPF shell for my Blazor app, it seems the most logical decision at the moment.

**Comment by robloo** [comment]
Note that a lot of this work is just implementing a WebView for Avalonia that works cross platform. This is definitely something Avalonia needs as a 1st party control in some form to stay competitive. I'm surprised there hasn't been more concern that it is missing.

This means I don't necessarily agree that BlazorWebView shouldn't be a "first class control in Avalonia". WebView must be a first class control sooner rather than later. Then once that is done, making a BlazorWebView is quite easy I imagine.

When I have time I will be happy to assist in starting a WebView control (@maxkatz6 already did quite some work). It is something that will be needed on my end as well. For mobile platforms it must wrap the native WebViews so we might as well do the same for Desktop. This means it will be much easier to maintain as we don't have to worry about all the details or keeping it up to the latest standards. This will likely be the only control that follows a MAUI-type architecture though.

**Reply by Laftek** [reply]
Completely agree. Blazor becoming respected in webdev world can only help Avalonia to get bigger audience and possible replacement for MAUI.

**Reply by anna-dolbina** [reply]
If you're interested in proprietary solutions, you can try DotNetBrowser, which actually provides a complete web view for Windows, Linux, and macOS:
https://avaloniaui.net/Blog/dotnetbrowser-meets-avalonia

It will work with Blazor too, btw, after certain integration efforts.

**Reply by vlad-lubenskyi** [reply]
We created a template for Blazor Hybrid & Avalonia UI apps, using DotNetBrowser as a web view:

```
dotnet new install DotNetBrowser.Templates
dotnet new dotnetbrowser.blazor.avalonia.app -o Blazor.AvaloniaUi -li <your_license_key>
dotnet run --project Blazor.AvaloniaUi
```

DotNetBrowser is not free, of course. But [here is an article](https://teamdev.com/dotnetbrowser/blog/blazor-hybrid-apps-in-avalonia-ui/) that explains how we integrated it all together, and replicating the approach with with anything CEF-based should be a no-brainer.

**Reply by AndrewPlakhotnyi** [reply]
12к$ per year for a control. Nice open source project :)

**Comment by trashhacker** [comment]
your could do it for windows with that https://github.com/AvaloniaUI/Avalonia/discussions/17185 or https://github.com/MicroSugarDeveloperOrg/Avalonia.WebView
and https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-components-outside-of-aspnetcore?view=aspnetcore-8.0

**Reply by AndrewPlakhotnyi** [reply]
This project is based on Xamarin.Mac, which reached end of support on May 1, 2024: https://github.com/dotnet/maui/discussions/21214

**Comment by sakib1361** [comment]
What happened to micro MicroSugarDeveloperOrg/Avalonia.WebView 
https://github.com/MicroSugarDeveloperOrg/Avalonia.WebView? 

Seems it got removed and there is no alternative. it worked mostly ok for windows and mac.

**Comment by HenKun** [comment]
This topic seems to be quite interesting and valuable. Is there any progress on that in 2025?

**Reply by icavalheiro** [reply]
Im also waiting for any news on this, but it seems not to be in any roadmap for the avalonia team :(

**Reply by maxkatz6** [reply]
Since this question was asked, we now have official [NativeWebView](https://docs.avaloniaui.net/accelerate/components/webview/quickstart) control.
But there are no plans for Blazor/Hybrid integration at the moment.

**Reply by mondermurshed** [reply]
More than three years and they still didn't listen to the community.

I will just use WPF as my shell at the moment.

**Reply by MikeCodesDotNET** [reply]
> More than three years and they still didn’t listen to the community.

Max explained three years ago that this is not on our roadmap and encouraged the community to build a solution instead. Our roadmap already represents multiple years of engineering effort, and we think very carefully about what should be a first party feature and what is better owned as a community project.

He said:

> if community wants to implement that and maintain as an third party library, we can help with information how it should be implemented.

Listening to the community does not always mean saying yes to every request. In the three years since, nobody has stepped up to design, implement and maintain a Blazor Hybrid integration, and we have not received a single PR adding it to core. That is usually a strong signal that interest is limited compared to everything else we are working on.

We are happy to support serious community efforts in this area with guidance and review, but we are not going to throw out a carefully planned roadmap and long term vision in order to chase whichever request happens to be shouted about the loudest.

**Reply by mondermurshed** [reply]
@MikeCodesDotNET understand and respect that Avalonia has a long-term roadmap and limited engineering resources, and I appreciate the transparency behind that.
However, I believe a Blazor Hybrid integration could bring significant value to the Avalonia ecosystem.

There is a large number of developers currently using Blazor Hybrid with .NET MAUI, and many of them would be very interested in trying Avalonia if first-party support, or even officially endorsed guidance, were available. This could meaningfully increase Avalonia’s adoption, especially among web-focused .NET developers who want a cross-platform desktop solution without committing to MAUI.

I also don’t fully agree that there has been little interest. Many developers assume that if something is not on the official roadmap, it’s not worth requesting or expecting, so they never push further. A clearer signal from the Avalonia team that community contributions in this area are encouraged might help.

In any case, I appreciate your openness to reviewing serious community efforts. Even an official design document, starting point, or minimal guidance from the team could help motivate contributors to step in and build the integration.

**Reply by icavalheiro** [reply]
Recently avalonia team posted on their blog about bringing maui to web and linux, I wonder how they would do that without support maui blazor hybrid, that would require avalonia to opensource their webview... Its clear that microsoft is trying to push blazor hybrid as a first class citizen on its maui platform, so i wonder what will happen when that linux layer comes out next year (or would that also be a payed feature just like the avalonia wpf thing?)

**Reply by tanshanli** [reply]
> This topic seems to be quite interesting and valuable. Is there any progress on that in 2025?

My current app is built with MAUI Blazor Hybrid. I was originally planning to migrate to the Avalonia platform because of its support for Linux and the web, but the need isn't strong enough at the moment. Moreover, I found that the official Avalonia team hasn't provided Blazor integration, so I've decided against making the switch.

**Reply by cyberdarius** [reply]
We do the same for many LOB applications: 

We have a blazor UI library that contains 90% of the app  (UI.MyApp)

and we incapsulate it in the required host: 

MyApp.MAUI -> Android, iOS
MyApp.WinDesktop -> WPF AddWpfBlazorWebView
MyApp.MacLinuxDesktop -> Photino.Blazor

MyApp.WebApp -> BlazorWebApp Template
MyApp.PWA -> FULL BlazorWASM

Electron/Tauri is a no go for desktop and mobile: we don't want to include complete browser.

ATM we must rely on MAUI, Photino and WPF but now with the official opensource WebView, an AvaloniaBlazorWebView on top of that would be great.

**Comment by jairbubbles** [comment]
Avalonia announced recently that their webview is going open source which is good news for this topic! 🥳 (See https://avaloniaui.net/blog/the-avalonia-webview-is-going-open-source/)

Based on https://github.com/tryphotino/photino.Blazor I was able to have Blazor Hybrid content inside an Avalonia application.

It would need some adaptation to make it work on all platforms as it's using WebView2 only on Windows. (https://docs.avaloniaui.net/docs/app-development/embedding-web-content#platform-prerequisites)

It's highly experimental but I'm sharing the code in case it can be useful to many more people:
[Avalonia.Blazor.zip](https://github.com/user-attachments/files/27472155/Avalonia.Blazor.zip)

I'm also trying to use https://github.com/wieslawsoltes/NativeWebVIew but didn't have time to work more on this. @wieslawsoltes I don't know if you have interest for this but having support for Blazor Hybrid could maybe be part of your repository?

**Reply by Denny09310** [reply]
I’ve built something similar, inspired by `Microsoft.AspNetCore.Components.WebView.Wpf`. Here’s the repository, including a sample project, in case anyone wants to extend it into a fully cross-platform solution. At the moment, I can confirm it works on Windows since it’s based on the WPF source code.

[AvaloniaApplication1.zip](https://github.com/user-attachments/files/27590409/AvaloniaApplication1.zip)
